doc/htmldoc/conf.py

```diff
@@ -247,6 +248,80 @@ def config_inited_handler(app, config):
     )
 
 
+def add_button_to_examples(app, env, docnames):
+    """Find all examples and include a link to launch notebook.
+
+     Function finds all restructured text files in auto_examples
+     and injects the multistring prolog, which is rendered
+     as a button link in HTML. The target is set to a Jupyter notebook of
+     the same name and a service to run it.
+     The nameholder in the string is replaced with the file name.
+
+     The rst files are generated at build time by Sphinx_gallery.
+     The notebooks that the target points to are linked with
+     services (like EBRAINS JupyterHub) that runs notebooks using nbgitpuller.
+     See https://hub.jupyter.org/nbgitpuller/link.html
+     The notebooks are located in the repository nest/nest-simulator-examples/.
+     The notebooks are generated from the CI workflow of NEST
+     on GitHub, which converts the source Python files to .ipynb.
+
+     The link to run the notebook is rendered in an image within a card directive.
+    """
+    example_prolog = """
+.. only:: html
+
+  .. card:: Run this example as a Jupyter notebook
+    :margin: auto
+    :width: 50%
+    :text-align: center
+
+    .. image:: https://nest-simulator.org/TryItOnEBRAINS.png
+         :target: https://lab.ebrains.eu/hub/user-redirect/git-pull?repo=\
+https%3A%2F%2Fgithub.com%2Fnest%2Fnest-simulator-examples\
+&urlpath=lab%2Ftree%2Fnest-simulator-examples%2Fnotebooks%2F\
+notebooks%2Ffilepath.ipynb&branch=main
+
+    For details and troubleshooting see :ref:`run_jupyter`."""
+
+    # Find all relevant files
+    # Inject prolog into Python example
+    files = list(Path(doc_build_dir / "auto_examples/").rglob("*.rst"))
+    for file in files:
+
+        # Skip index files and benchmark file. These files do not have notebooks that can run
+        # on the service.
+        if file.stem == "index" or file.stem == "hpc_benchmark":
+            continue
+
+        get_path = file.parent
+        get_subdir = get_path.stem
+
+        with open(file, "r") as f:
+
+            lines = f.readlines()
+
+            # get name of file and subdirectory to replace in link target
+            if get_subdir != "auto_examples":
+                path2example = get_subdir + "%2F" + file.stem
+                prolog = example_prolog.replace("filepath", path2example)
+
+            else:
+                path2example = file.stem
+                prolog = example_prolog.replace("filepath", path2example)
+
+        # find the first heading of the file.
+        for i, item in enumerate(lines):
+            if item.startswith("-----"):
+                break
+
+        # insert prolog into rst file after heading
+        lines.insert(i + 1, prolog + '\n')
+
+        with open(file, 'w') as f:
+            lines = "".join(lines)
+            f.write(lines)
+
```

Comment:
We don't need to open the file twice: once for reading and once for writing: I would do the following:

```python
with open(file, "r+") as f:
    data = f.readlines()
    data.modify()  # modify the line you want
    data = "".join(data)
    f.truncate(0) # clear everything from the file
    f.seek(0) # set pointer to the beginning of the file
    f.write(data)
```

Reply:
@Helveg, @jougs & @terhorstd: Any opinions regarding the use of `truncate` and `seek` instead of opening the file twice?

Reply:
I'd say it's premature optimization at best. Are we on the clock? :)

Reply:
Then @jessica-mitchell is up to decide which variant to use :p

Reply:
I went with back with the previous read/ write since it is what I'm familiar with.